portability-core/src/main/java/org/dataportabilityproject/shared/CloudAppCredentialFactory.java

```diff
@@ -38,24 +55,64 @@
       + "keyRings/portability_secrets/cryptoKeys/portability_secrets_key";
 
   private final CloudFactory cloudFactory;
+  private final BucketStore bucketStore;
+  private final CryptoKeyManagementSystem cryptoKeyManagementSystem;
 
   @Inject
   CloudAppCredentialFactory(CloudFactory cloudFactory) {
     this.cloudFactory = cloudFactory;
+    this.bucketStore = cloudFactory.getBucketStore();
+    this.cryptoKeyManagementSystem = cloudFactory.getCryptoKeyManagementSystem();
+    this.keys = CacheBuilder.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).build(
+        new CacheLoader<String, String>() {
+          @Override
+          public String load(String key) throws Exception {
+            return lookupKey(key);
+          }
+        });
+    this.secrets = CacheBuilder.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).build(
+        new CacheLoader<String, String>() {
+          @Override
+          public String load(String key) throws Exception {
+            return lookupSecret(key);
+          }
+        });
   }
 
   @Override
   public AppCredentials lookupAndCreate(String keyName, String secretName) throws IOException {
+    String key;
+    String secret;
+    try {
+      key = keys.get(keyName);
+    } catch (ExecutionException e) {
+      throw new IOException("Couldn't lookup key: " + keyName, e);
+    }
+
+    try {
+      secret = secrets.get(secretName);
+    } catch (ExecutionException e) {
+      throw new IOException("Couldn't lookup secret: " + secretName, e);
+    }
+
+    return AppCredentials.create(key, secret);
+  }
+
+  private String lookupKey(String keyName) {
     String projectId = cloudFactory.getProjectId();
     String bucketName = BUCKET_NAME_PREFIX + projectId;
-    BucketStore bucketStore = cloudFactory.getBucketStore();
-    CryptoKeyManagementSystem keyManagementSystem = cloudFactory.getCryptoKeyManagementSystem();
-
     String keyLocation = KEYS_DIR + keyName + KEY_EXTENSION;
     logger.debug("Getting key {} (blob {}) from bucket {}", keyName, keyLocation, bucketName);
     byte[] rawKeyBytes = bucketStore.getBlob(bucketName, keyLocation);
     String key = new String(rawKeyBytes);
     checkState(!Strings.isNullOrEmpty(key), "Couldn't lookup: " + keyName);
+    return key;
+  }
+
+  private String lookupSecret(String secretName) throws IOException {
+    String projectId = cloudFactory.getProjectId();
+    String bucketName = BUCKET_NAME_PREFIX + projectId;
```

Comment:
Could we do this without project id and just have bucketstore internally know the name of the bucket that way each cloud bucketstore doesn't have to follow a specific convention at this layer?

Reply:
I thought about this too but thought it could be good to have a convention for bucket names if it makes sense for other providers. Can we leave as is for now, and if another cloud provider comes along with a different convention, we can refactor to be impl specific?

Reply:
It just seems to me not imposing a convention  makes it easier than having another rule for a cloud provider or developer to follow. It seems the easiest onramp is to have the script create or takein a single flag, convention free.

Reply:
Done, see my reply to your other comment re crypto key